app/services/force/soql/preference_pagination_service.rb

```diff
@@ -0,0 +1,122 @@
+        # Given a comma-separated list of search terms:
+        # For each term, we search across name and application number for a match
+        # For multiple terms, we expect the record to match each of the terms individually.
+        search_terms = search_terms_string.split(',')
+        search_fields = [
+          'Application__r.Name',
+          'Application__r.Applicant__r.First_Name__c',
+          'Application__r.Applicant__r.Last_Name__c',
+        ]
+        search = []
+        search_terms.each do |search_term|
+          # For each term we join the fields with OR to return a record if it matches
+          # across any of these fields.
+          search_term_clause = []
+          sanitized_term = ActiveRecord::Base.sanitize_sql_like search_term
+          search_fields.each do |search_field|
+            search_term_clause.push("#{search_field} like '%#{sanitized_term}%'")
+          end
+          search.push("(#{search_term_clause.join(' OR ')})")
+        end
+        # If there are multiple terms, join the clauses with "AND" because we expect each term to find a match.
+        search.join(' AND ')
+      end
+
+      private
+
+      def app_preferences_for_listing_query(opts)
+        # The query for this was put together to combine the general
+        # preferences with all other preferences as general was not
+        # originally included as a preference name. Because of the
+        # complexity of the query we are dynamically adding filters
+        # within the WHERE clause with the helper function above. The
+        # Preference_All_Name and Preference_All_Lottery_Rank fields
+        # were added to bring all preferences into the same query.
+        filters = build_app_preferences_filters(opts)
+        search = opts[:search] ? build_app_preferences_search(opts[:search]) : nil
+        builder.from(:Application_Preference__c)
+               .select(query_fields(:app_preferences_for_listing))
+               .where(%(
+                Listing_ID__c = '#{opts[:listing_id].length >= 18 ? opts[:listing_id][0...-3] : opts[:listing_id]}'
+                #{search ? "AND (#{search})" : ''}
+                #{filters}
+                and \(\(Preference_Lottery_Rank__c != null and Receives_Preference__c = true\)
+                or \(Preference_Name__c = 'Live or Work in San Francisco Preference' and Application__r.General_Lottery_Rank__c != null\)\)
+                and Application__c IN \(SELECT id FROM Application__c\)
+              ))
+               .paginate(opts)
+               .order_by('Receives_Preference__c desc, Preference_Order__c, Preference_Lottery_Rank__c,  Application__r.General_Lottery_Rank__c asc')
+               .transform_results { |results| massage(results) }
+      end
+
+      def listing_subquery(listing_id)
+        builder.from(:Listing_Lottery_Preference__c)
+               .select(:Id)
+               .where_eq('Listing__c', "'#{listing_id}'")
+      end
+    end
+  end
+end
```

Comment:
question: Since you've created a new service file, are there plans to eventually remove the old `preferebnce_service.rb` file?